internal/engine/upper_test.go

```diff
@@ -340,9 +345,9 @@ func newFakeBuildAndDeployer(t *testing.T) *fakeBuildAndDeployer {
 func (b *fakeBuildAndDeployer) completeBuild(index int) {
 	ch, ok := b.getBuildCompletionChan(index)
 	if !ok {
-		// TODO(maia): maybe this should just create and close a channel for this key
-		b.t.Fatalf("tried to mark build #%d for for completion, but no completion channel registered (b.registerBuild(index) "+
-			"should already have been called--probably by b.BuildAndDeploy)", index)
+		// If build chan doesn't exist, create and store it
```

Comment:
this will cause race conditions. in async programming, you usually want to use LoadOrStore to atomically create something that doesn't exist
https://golang.org/pkg/sync/#Map.LoadOrStore

Reply:
i would probably change getBuildCompletionChan to getOrCreateBuildCompletionChan